src/App.tsx

```diff
@@ -1,26 +1,304 @@
+
+      return () => clearTimeout(timer);
+    }
+
+    return;
+  }, [error]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  function handleKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
+    setError('');
+    if (event.key !== 'Enter') {
+      return;
+    }
+
+    event.preventDefault();
+    if (inputTodo.trim() === '') {
+      setError('Title should not be empty');
+    } else {
+      setLoading(true);
+      setTodos(currentTodos => [
+        ...currentTodos,
+        {
+          id: Infinity,
+          title: inputTodo,
+          userId: USER_ID,
+          completed: false,
+        },
+      ]);
+      createTodo({
+        title: inputTodo.trim(),
+        userId: USER_ID,
+        completed: false,
+      })
+        .then(newTodo => {
+          setTodos(currentTodos => {
+            currentTodos.pop();
+
+            return [...currentTodos, newTodo];
+          });
+        })
+        .then(() => setInputTodo(''))
+        .catch(() => {
+          setTodos(currentTodo => {
+            currentTodo.pop();
+
+            return [...currentTodo];
+          });
+          setError('Unable to add a todo');
+        })
+        .finally(() => {
+          setLoading(false);
+        });
+    }
+  }
+
+  function handleEveryCompleted(todosSelect: Todo[]) {
+    setError('');
+    setLoading(true);
+    if (todosSelect.every(el => el.completed)) {
+      todosSelect.map(todo => {
+        updateTodo({
+          id: todo.id,
+          userId: USER_ID,
+          title: todo.title,
+          completed: !todo.completed,
+        })
+          .then(updatedTodo => {
+            setTodos(
+              [...todosSelect].map(todoRes => {
+                return { ...todoRes, completed: updatedTodo.completed };
+              }),
+            );
+          })
+          .catch(() => setError('Unable to update a todo'))
+          .finally(() => setLoading(false));
+      });
+    } else {
+      todosSelect
+        .filter(el => el.completed === false)
+        .map(todo => {
+          updateTodo({
+            id: todo.id,
+            userId: USER_ID,
+            title: todo.title,
+            completed: !todo.completed,
+          })
+            .then(updatedTodo => {
+              setTodos(
+                [...todosSelect].map(todoRes => {
+                  return { ...todoRes, completed: updatedTodo.completed };
+                }),
+              );
+            })
+            .catch(() => setError('Unable to update a todo'))
+            .finally(() => setLoading(false));
+        });
+    }
+  }
+
+  function handleDeleteCompletedTodos() {
+    const deletedID: number[] = [];
+
+    setError('');
+    setLoading(true);
+    todos
+      .filter(el => el.completed)
+      .forEach(todoCompleted => {
+        deleteTodo(todoCompleted.id)
+          .then(() => {
+            deletedID.push(todoCompleted.id);
+
+            setTodos(tds => tds.filter(el => !deletedID.includes(el.id)));
```

Comment:
TIP: When using a callback inside the `setState`, `tds` will already have the updated value, meaning every other item in `deletedID` will already have been deleted; take care of deleting only the current item:
`tds => tds.filter(el => el.id !== todoCompleted.id)`

Reply:
great